File: demo-api/app/api/v1/endpoints.py
import uuid
from datetime import datetime, timezone
from typing import List, Optional
from fastapi import APIRouter, HTTPException, Query
from app.models.catalog import (
    CartRequest, CartResponse, CheckoutRequest, CheckoutResponse,
    LoginRequest, LoginResponse, Product
)

router = APIRouter(tags=["E-Commerce API"])

# In-memory realistic product catalog
SAMPLE_PRODUCTS: List[Product] = [
    Product(id="prod-001", name="Ultra-Shield Cloud WAF", category="security", price=299.99, inventory=150, description="Enterprise edge security layer."),
    Product(id="prod-002", name="Neural Scale Pod", category="compute", price=49.99, inventory=500, description="Auto-optimizing container pod allocation."),
    Product(id="prod-003", name="TraceIQ Observability Agent", category="telemetry", price=89.50, inventory=250, description="Distributed tracing and span analytics."),
    Product(id="prod-004", name="RateGuard Rate Limiter", category="security", price=120.00, inventory=80, description="Token-bucket rate limiter with IP reputation."),
    Product(id="prod-005", name="KubeBalance Service Mesh", category="networking", price=199.00, inventory=120, description="Traffic routing and mTLS infrastructure."),
]


@router.get("/products", response_model=List[Product])
async def list_products(
    category: Optional[str] = None,
    limit: int = Query(default=10, ge=1, le=50)
):
    """Retrieve catalog of products."""
    products = SAMPLE_PRODUCTS
    if category:
        products = [p for p in products if p.category.lower() == category.lower()]
    return products[:limit]


@router.get("/products/{product_id}", response_model=Product)
async def get_product(product_id: str):
    """Retrieve single product by ID."""
    for p in SAMPLE_PRODUCTS:
        if p.id == product_id:
            return p
    raise HTTPException(status_code=404, detail="Product not found")


@router.get("/search", response_model=List[Product])
async def search_products(q: str = Query(..., min_length=1)):
    """Search products by keyword."""
    query = q.lower()
    return [p for p in SAMPLE_PRODUCTS if query in p.name.lower() or query in p.description.lower()]


@router.post("/login", response_model=LoginResponse)
async def login(request: LoginRequest):
    """Simulate user authentication."""
    if not request.username or not request.password:
        raise HTTPException(status_code=400, detail="Username and password required")
    return LoginResponse(
        token=f"jwt-mock-{uuid.uuid4().hex}",
        user_id=f"user-{request.username}",
        expires_in_seconds=3600
    )


@router.post("/cart", response_model=CartResponse)
async def update_cart(request: CartRequest):
    """Add items to user cart."""
    product_map = {p.id: p.price for p in SAMPLE_PRODUCTS}
    total = sum(product_map.get(item.product_id, 50.0) * item.quantity for item in request.items)
    return CartResponse(
        cart_id=f"cart-{uuid.uuid4().hex[:8]}",
        user_id=request.user_id,
        items=request.items,
        total_amount=round(total, 2)
    )


@router.post("/checkout", response_model=CheckoutResponse)
async def checkout(request: CheckoutRequest):
    """Process shopping cart checkout."""
    return CheckoutResponse(
        order_id=f"order-{uuid.uuid4().hex[:10]}",
        cart_id=request.cart_id,
        status="completed",
        charged_amount=349.98,
        timestamp=datetime.now(timezone.utc).isoformat()
    )

File: load-tests/k6/endpoints.js
/**
 * SentinelScale k6 Load Test - Demo API Endpoints Module
 *
 * Implements realistic API interaction helpers representing genuine user journeys:
 * - Product browsing & catalog pagination
 * - Keyword search
 * - Single item product detail retrieval
 * - User authentication (login)
 * - Cart updates
 * - Checkout processing
 * - Health / ready probes
 */

import http from 'k6/http';
import { check } from 'k6';

export const SAMPLE_PRODUCT_IDS = [
  'prod-001',
  'prod-002',
  'prod-003',
  'prod-004',
  'prod-005',
];

export const SEARCH_KEYWORDS = [
  'security',
  'compute',
  'telemetry',
  'pod',
  'waf',
  'rate',
  'mesh',
];

export const CATEGORIES = [
  'security',
  'compute',
  'telemetry',
  'networking',
];

const JSON_HEADERS = {
  'Content-Type': 'application/json',
  'Accept': 'application/json',
};

/**
 * Health check probe
 */
export function checkHealth(baseUrl) {
  const res = http.get(`${baseUrl}/health`, { headers: JSON_HEADERS, tags: { endpoint: 'health' } });
  return check(res, {
    'health status is 200': (r) => r.status === 200,
  });
}

/**
 * Browse products catalog
 */
export function listProducts(baseUrl, category = null, limit = 10) {
  let url = `${baseUrl}/products?limit=${limit}`;
  if (category) {
    url += `&category=${category}`;
  }
  const res = http.get(url, { headers: JSON_HEADERS, tags: { endpoint: 'products_list' } });
  return check(res, {
    'products list status is 200': (r) => r.status === 200,
    'products list returns array': (r) => {
      try {
        const body = JSON.parse(r.body);
        return Array.isArray(body);
      } catch (e) {
        return false;
      }
    },
  });
}

/**
 * Get product detail by ID
 */
export function getProduct(baseUrl, productId) {
  const res = http.get(`${baseUrl}/products/${productId}`, { headers: JSON_HEADERS, tags: { endpoint: 'product_detail' } });
  return check(res, {
    'product detail status is 200': (r) => r.status === 200,
    'product detail has id': (r) => {
      try {
        const body = JSON.parse(r.body);
        return body.id === productId;
      } catch (e) {
        return false;
      }
    },
  });
}

/**
 * Search products by keyword
 */
export function searchProducts(baseUrl, query) {
  const res = http.get(`${baseUrl}/search?q=${encodeURIComponent(query)}`, { headers: JSON_HEADERS, tags: { endpoint: 'search' } });
  return check(res, {
    'search status is 200': (r) => r.status === 200,
    'search returns array': (r) => {
      try {
        const body = JSON.parse(r.body);
        return Array.isArray(body);
      } catch (e) {
        return false;
      }
    },
  });
}

/**
 * Authenticate user
 */
export function loginUser(baseUrl, username = 'demo_user', password = 'password123') {
  const payload = JSON.stringify({ username, password });
  const res = http.post(`${baseUrl}/login`, payload, { headers: JSON_HEADERS, tags: { endpoint: 'login' } });
  return check(res, {
    'login status is 200': (r) => r.status === 200,
    'login returns token': (r) => {
      try {
        const body = JSON.parse(r.body);
        return !!body.token;
      } catch (e) {
        return false;
      }
    },
  });
}

/**
 * Add item to cart
 */
export function updateCart(baseUrl, userId, productId, quantity = 1) {
  const payload = JSON.stringify({
    user_id: userId,
    items: [{ product_id: productId, quantity: quantity }],
  });
  const res = http.post(`${baseUrl}/cart`, payload, { headers: JSON_HEADERS, tags: { endpoint: 'cart' } });
  return check(res, {
    'cart status is 200': (r) => r.status === 200,
    'cart returns cart_id': (r) => {
      try {
        const body = JSON.parse(r.body);
        return !!body.cart_id;
      } catch (e) {
        return false;
      }
    },
  });
}

/**
 * Process order checkout
 */
export function checkout(baseUrl, cartId = 'cart-test-01') {
  const payload = JSON.stringify({
    cart_id: cartId,
    payment_method: 'credit_card',
    shipping_address: '100 Enterprise Way, Suite 400, Cloud City, CC 94016',
  });
  const res = http.post(`${baseUrl}/checkout`, payload, { headers: JSON_HEADERS, tags: { endpoint: 'checkout' } });
  return check(res, {
    'checkout status is 200': (r) => r.status === 200,
    'checkout status is completed': (r) => {
      try {
        const body = JSON.parse(r.body);
        return body.status === 'completed';
      } catch (e) {
        return false;
      }
    },
  });
}
